io.c

```diff
@@ -2685,17 +2699,46 @@ static int io_uring_release(struct inode *inode, struct file *file)
 	return 0;
 }
 
+static long io_ring_ioctl_init(struct io_ring_ctx *ctx, unsigned long arg)
+{
+	struct io_uring_params params;
+	int ret;
+
+	if (copy_from_user(&params, (void *)arg, sizeof(params)))
+		return -EFAULT;
+
+	ret = io_ring_ctx_init(ctx, &params);
+	if (ret != 0)
+		return ret;
+
+	ret = io_sq_offload_start(ctx, &params);
+	if (ret != 0) {
+		io_ring_ctx_wait_and_kill(ctx);
+		return ret;
+	}
+
+	if (copy_to_user((void *)arg, &params, sizeof(params)))
+		return -EFAULT;
+
+	return 0;
+}
+
 static long io_uring_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 {
 	struct io_ring_ctx *ctx = filp->private_data;
 
 	switch (cmd) {
+	case PXD_IOC_WAKE_UP_SQO:
+		wake_up(&ctx->sqo_wait);
+		return 0;
 	case PXD_IOC_RUN_IO_QUEUE:
 		return io_run_queue(ctx);
 	case PXD_IOC_REGISTER_FILE:
 		return io_sqe_register_file(ctx, arg);
 	case PXD_IOC_UNREGISTER_FILE:
 		return io_sqe_unregister_file(ctx, arg);
+	case PXD_IOC_INIT_IO:
+		return io_ring_ioctl_init(ctx, arg);
```

Comment:
Why the new ioctls, and why the move out of io_uring_open() ?

Reply:
these are the missed changes.
io uring now has additional params through which the queue depth etc can be setup.
So there will be an initial ioctl INIT_IO to set this up.
And then usage will be as usual. Earlier it was single step with default hardcoded params.